### DISCUSSION TITLE
how to declare a generic `service handler`?

### DISCUSSION THREAD

**Discussion by jobenye** [hyperium/hyper#3141] [Q&A]
version: 1.0.0-rc.2

I always failed to declare a generic `service handler`, would anyone please give a help?
Here is my code
```
pub async fn generic_handler(req: Request<Incoming>) -> Result<Response<impl Body>, Infallible>
{
    let x = Full::new(Bytes::from("Hello world"));

    Ok(Response::new(x))
}
```

After passing it to `.serve_connection(stream, service_fn(handler::generic_handler))`, I got a lot of errors,
e.g.
the trait `StdError` is not implemented for `<impl Body as Body>::Error`
`<impl Body as Body>::Error` cannot be sent between threads safely
...

====

Another question is, is that possible to handle both http1 and http2 in one `Builder`, i.e. `http::Builder::new()....`. As to one `service handler` and one port can serve either http1 or http2. Or are there any other ways except for checking the ALPN before running the `server_connection()`?

**Comment by seanmonstar** [comment]
You need to specify the associated types of `Body`, like `Response<impl Body<Data = impl Buf, Error = Infallible>>`. You might need to add a little more information, just follow the compiler errors, they give instructions on what else it wants to know.

**Reply by jobenye** [reply]
Thanks for the response.

Sorry I didn't make my question clear. My purpose is as below, not sure if it is possible?

```
pub async fn generic_handler(req: Request<Incoming>) -> Result<Response<impl Body>, Infallible>
{
    if url == "path1" {
         Ok(Response::new(Full::new(Bytes::from("Hello world"))))
    } else if url == "path2" {
        Ok(Response::new(http_body_util::StreamBody::new(some_stream)));
    } else {
        ...
    }
}
```

**Reply by seanmonstar** [reply]
This is a general Rust thing: in order to return different types, you need to unify them somehow. Generally, that either means using an `enum`, or a trait object. You could either use `http_body_util::Either`, or `http_body::combinators::BoxBody`.

**Reply by jobenye** [reply]
> This is a general Rust thing: in order to return different types, you need to unify them somehow. Generally, that either means using an `enum`, or a trait object. You could either use `http_body_util::Either`, or `http_body::combinators::BoxBody`.

Get it, thanks a lot.

**Reply by jobenye** [reply]
> This is a general Rust thing: in order to return different types, you need to unify them somehow. Generally, that either means using an `enum`, or a trait object. You could either use `http_body_util::Either`, or `http_body::combinators::BoxBody`.

Eventually, I got the following solution. Write here in case someone else has similar demands.

```
type MyBody = Pin<Box<dyn Body<Data = Box<dyn Buf + Send>, Error = Box<dyn StdError + Send + Sync>> + Send>>;
pub async fn generic_handler(req: Request<Incoming>) -> Result<Response<MyBody>, Error>;
```

Don't know how to close this question ...
